### PR TITLE
feat(auth): automatically refresh tokens

### PR DESCRIPTION
**Pull request by Chris-Bitler**
Automatically refresh the user authentication tokens on a 30 minute interval to make sure we don't start getting 401's for the store requests

### CODE REVIEW

**Comment by danreeves** [comment]
This looks good, thanks for the PR. I've been wondering if this works or is better included in the background.js file.

File: src/utils.ts [hunk 1]
```diff
@@ -59,6 +59,11 @@ export function getLocalStorage<T>(key: string): T | undefined {
 	return decoded
 }
 
+export function setLocalStorage<T>(key: string, value: T) {
+	const encoded = btoa(JSON.stringify(value));
+	localStorage.setItem(key, encoded);
+}
```
Comment (danreeves) [inline src/utils.ts:56]: The official site code uses this:

```ts
function setLocalStorage<T>(key: string, data: T): void {
    const encoded = Buffer.from(JSON.stringify(data), 'binary').toString('base64')
    localStorage.setItem(key, encoded)
}
```

Do you know if this has any difference?

File: src/hooks/useAuth.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { useEffect } from "react"
+import type { User } from "../types"
+import { createFetcher, getLocalStorage, setLocalStorage } from "../utils"
+
+const REFRESH_TIMER = 60 * 1000 * 30; // 30 minutes in ms
```
Comment (danreeves) [inline src/hooks/useAuth.ts]: The refresh response includes a ExpiresIn key, it would be cool to use that in case it changed to be shorter than 30 mins

File: src/utils.ts (hunk 1 above)
Comment (Chris-Bitler) [reply src/utils.ts:56]: As far as I'm aware, there shouldn't be any difference in terms of the resulting data - tested it by converting it back and forth and reloading the page so that the page's code would have to grab the newly stored value.

However, if you want it to match their implementation exactly I can update it

File: src/hooks/useAuth.ts (hunk 2 above)
Comment (Chris-Bitler) [reply src/hooks/useAuth.ts]: Will do

File: src/utils.ts (hunk 1 above)
Comment (Coconutcoo) [reply src/utils.ts:56]: To my knowledge, using Buffer.from over btoa is probably an artifact of the dev that wrote it being used to NodeJS. There Buffer is a global since btoa is a browser-only thing.

Comment (danreeves) [reply src/utils.ts:56]: Good to hear. I just didn't want to accidentally interfere with the sites normal operation

File: src/components/Layout.tsx [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import { useState } from "react"
+import { useEffect, useState } from "react"
```
Comment (Coconutcoo) [inline src/components/Layout.tsx]: Is this import needed? Can't see you using it here

**Review by Coconutcoo — COMMENTED** [review]
I think unneeded import, also, is it worth us trying to implement the 'pause' option a user asked at the same time here?

I think we can simply unset all the alarms, and have re-enabling it just reload?

File: src/utils.ts (hunk 1 above)
Comment (danreeves) [reply src/utils.ts:56]: I need to test some more but btoa actually errors if the string contains unicode https://developer.mozilla.org/en-US/docs/Web/API/btoa#unicode_strings

image

These kinds of names are pretty common in Steam accounts

File: src/background.ts [hunk 4]
```diff
@@ -1,6 +1,43 @@
...
 let ext = chrome || browser
 let dashboardUrl = "https://accounts.atoma.cloud/dashboard"
+let authData: User | null = null;
 
 ext.browserAction.onClicked.addListener(() => {
 	window.open(dashboardUrl, "_blank")
-})
+});
+
+const createTimer = () => ext.alarms.create("user-auth-refresh", { delayInMinutes: 1 })
```
Comment (danreeves) [inline src/background.ts]: So this runs every minute or so. I think it would be good to be more conservative and only update as much as necessary

File: src/background.ts [hunk 5]
```diff
@@ -1,6 +1,43 @@
...
+});
+
+ext.tabs.onRemoved.addListener((tabId) => {
+	atomaTabs = atomaTabs.filter((tab) => tab != tabId)
+});
+
+ext.alarms.onAlarm.addListener((alarm) => {
+	if(alarm.name == "user-auth-refresh") {
+		createTimer();
+		if (authData) {
```
Comment (danreeves) [inline src/background.ts]: Should we have a check if the auth isn't expiring soon then skip?

**Comment by Chris-Bitler** [comment]
@danreeves I got wrapped up in other stuff yesterday and forgot about making sure we only refresh it when necessary. Will fix it today

File: src/utils.ts (hunk 1 above)
Comment (Coconutcoo) [reply src/utils.ts:56]: Yeah we literally just realised this on Discord - may just want to use Buffers everywhere if we have to pay the 2k lines code in bundle.js anyway

File: src/background.ts (hunk 5 above)
Comment (Chris-Bitler) [reply src/background.ts]: Done

File: extension/manifest.json [hunk 6]
```diff
@@ -8,9 +8,8 @@
     "128": "icon128.png"
   },
   "permissions": [
-    "https://accounts.atoma.cloud/*",
-    "https://bsp-td-prod.atoma.cloud/*",
-    "https://bsp-cdn-prod.atoma.cloud/*"
```
Comment (danreeves) [inline extension/manifest.json]: I'd like to leave them explicitly here even if they're not necessary

File: extension/manifest.json [hunk 7]
```diff
@@ -8,9 +8,8 @@
...
+    "alarms",
+    "*://accounts.atoma.cloud/*"
```
Comment (danreeves) [inline extension/manifest.json]: Does this present as anything scary to users? like when the tabs permission requested browser history access

File: package.json [hunk 8]
```diff
@@ -21,23 +21,22 @@
     "swr": "^2.0.0"
   },
   "devDependencies": {
+    "@testing-library/react": "^13.4.0",
     "@tsconfig/strictest": "^1.0.2",
+    "@types/chrome": "^0.0.206",
```
Comment (danreeves) [inline package.json:26]: Nice

File: src/background.ts [hunk 9]
```diff
@@ -1,6 +1,55 @@
+import type { User } from "./types";
+import { createFetcher } from "./utils";
+
+// @ts-ignore
```
Comment (danreeves) [inline src/background.ts]: Can we add https://github.com/kelseasy/web-ext-types or define our own global.d.ts file that does something like `declare namespace browser = chrome` (pseudo code)

File: src/background.ts [hunk 10]
```diff
@@ -1,6 +1,55 @@
...
+		if (authData) {
+			try {
+				const fetcher = createFetcher(authData, true);
+				const result = await fetcher("https://bsp-auth-prod.atoma.cloud/queue/refresh")
+				
+				// Leave some time for error in the refresh
+				const expiresInMinutes = Math.floor((result.ExpiresIn ?? 1800) / 60) - 5;
+				createTimer(expiresInMinutes);
+				authData = result;
+				chrome.tabs.query({}, (tabs) => {
```
Comment (danreeves) [inline src/background.ts]: Doesn't querying tabs require tabs permission?

File: src/components/context/UserContextProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,45 @@
+// @ts-ignore
```
Comment (danreeves) [inline src/components/context/UserContextProvider.tsx]: What's being ignored? I'd rather we didn't ignore errors and if we do then provide a comment as to why

File: src/components/context/UserContextProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,45 @@
+// @ts-ignore
+import React, { createContext, useContext, useEffect, useState, ReactNode } from "react";
+import type { User } from "../../types";
+import { getFatSharkUser, setLocalStorage } from "../../utils"
+
+// @ts-ignore
```
Comment (danreeves) [inline src/components/context/UserContextProvider.tsx]: Same as above

File: src/types.ts [hunk 13]
```diff
@@ -223,3 +223,5 @@ export interface FilterRule {
 
 export const STORE_TYPES = ["credits", "marks"] as const
 export type StoreType = typeof STORE_TYPES[number]
+
+declare module browser {}
```
Comment (danreeves) [inline src/types.ts]: globals should go into a different file but this should get replaced with the types library i linked above

File: src/utils.ts (hunk 1 above)
Comment (danreeves) [reply src/utils.ts:56]: This still needs fixing. We should use the same code as the official site 

```ts
function setLocalStorage<T>(key: string, data: T): void {
    const encoded = Buffer.from(JSON.stringify(data), 'binary').toString('base64')
    localStorage.setItem(key, encoded)
}
```

File: tests/background.test.ts [hunk 14]
```diff
@@ -0,0 +1,99 @@
+// @ts-nocheck
```
Comment (danreeves) [inline tests/background.test.ts]: We can add jest types `@types/jest`

**Review by danreeves — CHANGES_REQUESTED** [review]
A few other, more general, things

1. The UserContext doesn't seem necessary. The whole "event listener syncing background and localstorage" thing can happen outside of react and then then useFetcher can go back to just getting the user from local storage.

2. I prefer test files being colocated with the thing they're testing, unless I'm outnumbered I'd like to keep that

3. Can you run prettier over the files you edit? `npx prettier src --write` should do it. Thinking about adding a precommit hook for this

File: extension/manifest.json (hunk 7 above)
Comment (Chris-Bitler) [reply extension/manifest.json]: @danreeves This is what it says - it shouldn't be nearly as scary to users and is better than the alternative
image

File: src/background.ts (hunk 10 above)
Comment (Chris-Bitler) [reply src/background.ts]: Not if it is a tab that we have host-specific permissions for, hence the accounts.atoma.cloud being in permissions

**Review by danreeves — COMMENTED** [review]
This is looking good. I'm going to leave a tab open for a couple hours to see what happens, then merge

**Comment by Coconutcoo** [comment]
@danreeves What was the outcome of you leaving it open for a bit?

**Comment by danreeves** [comment]
@Coconutcoo copy pasting some messages i put in discord:

i had it running in the background while i worked and sometimes it would get stuck on a loading spinner when i tabbed back. i think it's to do with the fetcher getting memoized. i did a little refactor on the branch locally so it pulls out of localStorage every time it tries to fetch and it appears to have resolved that.

opening the laptop again in the morning i got TypeError: Cannot read properties of undefined (reading 'ExpiresIn') in the background thread

**Comment by danreeves** [comment]
Closing as stale. Feel free to re-open a PR if you make any more progress on it!
